src/io_threads.c

```diff
@@ -574,6 +574,7 @@ void trySendPollJobToIOThreads(void) {
 static void ioThreadAccept(void *data) {
     client *c = (client *)data;
     connAccept(c->conn, NULL);
+    atomic_thread_fence(memory_order_release);
     c->io_read_state = CLIENT_COMPLETED_IO;
```

Comment:
Why not do the fence after updating `c->io_read_state` to make `processIOThreadsReadDone` see that it's completed ASAP?

Reply:
> Why not do the fence after updating `c->io_read_state` to make `processIOThreadsReadDone` see that it's completed ASAP? There's a risk the read state will be seen but not the flags... So why is `io_read_state` marked as volatile rather than atomic?

Reply:
`io_read_state` is not marked as atomic for performance reasons because we read it in many places in the main thread, mostly just as sanity checks. It is marked as volatile to avoid incorrect compilation time optimizations.
As we activate IO threads only under load, the possibility that it may take a bit longer  for the main thread to get the updated results is amortized and we didn't observe a degradation because of  that. 
In any way, we intend to use MPSC queue for the IO threads to signal the main thread when a job is done.

Reply:
> `io_read_state` is not marked as atomic for performance reasons because we read it in many places in the main thread, mostly just as sanity checks. It is marked as volatile to avoid incorrect compilation time optimizations.

Thanks, it could be useful to add a code comment about this in server.h at the declaration of these volatile variables. Otherwise I will forget it very soon and ask you again in another PR. ;)

> we intend to use MPSC queue for the IO threads to signal the main thread when a job is done

Interesting! Are you working on this now?